torch/csrc/generic/Tensor.cpp

```diff
@@ -498,14 +498,115 @@ static bool THPTensor_(_indexOnce)(PyObject *index, int &indexed_dim,
+
+  if (PySequence_Check(arg)) {
+    Py_ssize_t indexers = PySequence_Size(arg);
+    if (indexers == ndim) {
+      for (Py_ssize_t i = 0; i < indexers; ++i) {
+        PyObject *item = PySequence_GetItem(arg, i);
+        if (!THPIndexTensor_Check(item)) {
+          return false;
+        }
+      }
+      return true;
+    }
+  }
+  return false;
+
+  /**
+  // Checks whether the specified selection object should trigger advanced
+  // indexing
+
+  // Case 1: arg is a non-tuple sequence object
+  if (PySequence_Check(arg) && !PyTuple_Check(arg)) return true;
+
+#ifdef WITH_NUMPY
+  // Case 2: arg is an nd-array with type integer or bool
+  if (PyArray_Check(arg) && (PyArray_TYPE((PyArrayObject*)arg) == NPY_INT64 || PyArray_TYPE((PyArrayObject*)arg) == NPY_BOOL)) return true;
+#endif
+
+  // Case 3: arg is a tuple containing at least one sequence object or ndarray
+  if (PyTuple_Check(arg)) {
+    for (Py_ssize_t i = 0; i < PyTuple_GET_SIZE(arg); ++i) {
+      PyObject *item = PyTuple_GET_ITEM(arg, i);
+      if (PySequence_Check(item)) {
+        return true;
+      }
+      // TODO: add check for LongTensor?
+#ifdef WITH_NUMPY
+      if (PyArray_Check(item) && (PyArray_TYPE((PyArrayObject*)item) == NPY_INT64 || PyArray_TYPE((PyArrayObject*)item) == NPY_BOOL)) return true;
+#endif
+    }
+  }
+
+  return false;
+  **/
+}
+
+static bool THPTensor_(_advancedIndex)(
+    PyObject *index, THTensorPtr &tresult, THStorage * &sresult, long &storage_offset)
 {
   // Precondition: index is an object that specifies advanced indexing.
   // For now, we only support the simple integer-array indexing strategy
   // where there are ndim(self) indexing Long Tensors that can be broadcasted
   // and iterated as one
+  // TODO: empty indexer?
+
+  // First, verify that all of the indexers have the same shape, later we will
+  // incorporate broadcasting
+  // Also assuming for now they are all 1D, and not 2D, 3D, etc.
+  Py_ssize_t size = PySequence_Size(index);
+  THIndexTensor *first = ((THPIndexTensor *)PySequence_GetItem(index, 0))->cdata;
+  for (Py_ssize_t i = 1; i < size; ++i) {
+    if (!THIndexTensor_(isSameSizeAs)(LIBRARY_STATE first, ((THPIndexTensor *)PySequence_GetItem(index, i))->cdata)) {
+      PyErr_Format(PyExc_IndexError,
+          "When performing advanced indexing the shapes of the indexing Tensors must be the same");
+      return false;
+    }
+  }
+
+  // Our strategy is to view the indexed Tensor as a 1D Tensor, calculate
+  // the linear indices for each tuple of indexing elements, and then call
+  // indexSelect using those linear indices
+  // TODO: example
+
+  THIndexTensor *linearIndices = THIndexTensor_(newWithSize1d)(LIBRARY_STATE THIndexTensor_(nElement)(LIBRARY_STATE first));
+  for (ptrdiff_t i = 0; i < THIndexTensor_(nElement)(LIBRARY_STATE linearIndices); ++i) {
+    long stride = 1; // is this an incorrect assumption? could the stride at the last dim be non-one?
```

Comment:
If the tensor is not contiguous, then the stride of the last dimension might not be 1.